Tuples don't work as object attributes

```
[case testTupleAttr]
from typing import Tuple
class C:
    d: Tuple[int, object]
[file driver.py]
```
produces errors like
```
tmp/native.c: In function ‘native_C_getd’:
tmp/native.c:62:17: error: invalid operands to binary == (have ‘struct tuple_def_248584400443490’ and ‘struct tuple_def_248584400443490’)
     if (self->d == tuple_undefined_248584400443490) {
         ~~~~~~~ ^~
```

We need to generate different code for checking whether a tuple is undefined. Since tuple definedness is (I think) all or nothing, we can probably just (recursively) check the first attribute's definedness.

--- a/mypyc/emit.py
+++ b/mypyc/emit.py
@@ -8,7 +8,8 @@ from mypyc.ops import (
     Any, AssignmentTarget, Environment, BasicBlock, Value, Register, RType, RTuple, RInstance,
     ROptional, RPrimitive, is_int_rprimitive, is_float_rprimitive, is_bool_rprimitive,
     short_name, is_list_rprimitive, is_dict_rprimitive, is_set_rprimitive, is_tuple_rprimitive,
-    is_none_rprimitive, is_object_rprimitive, object_rprimitive, is_str_rprimitive, ClassIR, FuncIR
+    is_none_rprimitive, is_object_rprimitive, object_rprimitive, is_str_rprimitive, ClassIR,
+    FuncIR, int_rprimitive
 )
 from mypyc.namegen import NameGenerator
 
@@ -170,19 +171,27 @@ class Emitter:
 
         return result
 
-    def tuple_check_cond(
+    def tuple_undefined_check_cond(
             self, rtuple: RTuple, tuple_expr_in_c: str,
             c_compare_type_func: Callable[[RType], str], compare: str) -> str:
-        item_expr_in_c = tuple_expr_in_c + '.f0'
-        item_type = rtuple.types[0]
-        while isinstance(item_type, RTuple):
-            item_type = item_type.types[0]
-            item_expr_in_c += '.f0'
-        return '{} {} {}'.format(
-            item_expr_in_c,
-            compare,
-            c_compare_type_func(item_type)
-        )
+        # see tuple_c_declaration() above
+        if len(rtuple.types) == 0:
+            return '{} {} {}'.format(
+                tuple_expr_in_c + '.dummy_var_to_avoid_empty_struct',
+                compare,
+                c_compare_type_func(int_rprimitive)
+            )
+        else:
+            item_expr_in_c = tuple_expr_in_c + '.f0'
+            item_type = rtuple.types[0]
+            while isinstance(item_type, RTuple):
+                item_type = item_type.types[0]
+                item_expr_in_c += '.f0'
+            return '{} {} {}'.format(
+                item_expr_in_c,
+                compare,
+                c_compare_type_func(item_type)
+            )
 
     def tuple_undefined_value(self, rtuple: RTuple) -> str:
         context = self.context
@@ -197,6 +206,9 @@ class Emitter:
 
     def tuple_undefined_value_helper(self, rtuple: RTuple) -> List[str]:
         res = []
+        # see tuple_c_declaration()
+        if len(rtuple.types) == 0:
+            return [self.c_undefined_value(int_rprimitive)]
         for item in rtuple.types:
             if not isinstance(item, RTuple):
                 res.append(self.c_undefined_value(item))
@@ -491,7 +503,7 @@ class Emitter:
             if len(rtype.types) == 0:
                 return  # empty tuples can't fail.
             else:
-                cond = self.tuple_check_cond(rtype, value, self.c_error_value, '==')
+                cond = self.tuple_undefined_check_cond(rtype, value, self.c_error_value, '==')
                 self.emit_line('if ({}) {{'.format(cond))
         self.emit_lines(failure, '}')
 

--- a/mypyc/emit.py
+++ b/mypyc/emit.py
@@ -173,25 +173,18 @@ class Emitter:
 
     def tuple_undefined_check_cond(
             self, rtuple: RTuple, tuple_expr_in_c: str,
-            c_compare_type_func: Callable[[RType], str], compare: str) -> str:
-        # see tuple_c_declaration() above
-        if len(rtuple.types) == 0:
-            return '{} {} {}'.format(
-                tuple_expr_in_c + '.dummy_var_to_avoid_empty_struct',
-                compare,
-                c_compare_type_func(int_rprimitive)
-            )
+            c_type_compare_val: Callable[[RType], str], compare: str) -> str:
+        item_type = rtuple.types[0]
+        if not isinstance(item_type, RTuple):
+            return '{}.f0 {} {}'\
+                .format(tuple_expr_in_c, compare, c_type_compare_val(item_type))
+        elif isinstance(item_type, RTuple) and len(item_type.types) == 0:
+            # empty tuple
+            return '{}.dummy_var_to_avoid_empty_struct {} {}'\
+                .format(tuple_expr_in_c, compare, c_type_compare_val(int_rprimitive))
         else:
-            item_expr_in_c = tuple_expr_in_c + '.f0'
-            item_type = rtuple.types[0]
-            while isinstance(item_type, RTuple):
-                item_type = item_type.types[0]
-                item_expr_in_c += '.f0'
-            return '{} {} {}'.format(
-                item_expr_in_c,
-                compare,
-                c_compare_type_func(item_type)
-            )
+            return self.tuple_undefined_check_cond(
+                item_type, tuple_expr_in_c + '.f0', c_type_compare_val, compare)
 
     def tuple_undefined_value(self, rtuple: RTuple) -> str:
         context = self.context

--- a/mypyc/emit.py
+++ b/mypyc/emit.py
@@ -1,7 +1,7 @@
 """Utilities for emitting C code."""
 
 from collections import OrderedDict
-from typing import List, Set, Dict, Optional, List
+from typing import List, Set, Dict, Optional, List, Callable
 
 from mypyc.common import REG_PREFIX, STATIC_PREFIX, TYPE_PREFIX, NATIVE_PREFIX
 from mypyc.ops import (
@@ -169,6 +169,23 @@ class Emitter:
         result.append('')
 
         return result
+
+    def tuple_check_cond(
+            self,
+            rtuple: RTuple,
+            tuple_expr_in_c: str,
+            c_compare_type_func: Callable[[RType], str],
+            compare: str) -> str:
+        item_expr_in_c = tuple_expr_in_c + '.f0'
+        item_type = rtuple.types[0]
+        while isinstance(item_type, RTuple):
+            item_type = item_type.types[0]
+            item_expr_in_c += '.f0'
+        return '{} {} {}'.format(
+            item_expr_in_c,
+            compare,
+            c_compare_type_func(item_type)
+        )
 
     def tuple_undefined_value(self, rtuple: RTuple) -> str:
         context = self.context
@@ -477,8 +494,8 @@ class Emitter:
             if len(rtype.types) == 0:
                 return  # empty tuples can't fail.
             else:
-                self.emit_line('if ({}.f0 == {}) {{'.format(value,
-                                                            self.c_error_value(rtype.types[0])))
+                cond = self.tuple_check_cond(rtype, value, self.c_error_value, '==')
+                self.emit_line('if ({}) {{'.format(cond))
         self.emit_lines(failure, '}')
 
     def emit_gc_visit(self, target: str, rtype: RType) -> None:

--- a/mypyc/emit.py
+++ b/mypyc/emit.py
@@ -176,8 +176,15 @@ class Emitter:
         name = 'tuple_undefined_' + id
         if name not in context.statics:
             struct_name = self.tuple_struct_name(rtuple)
-            values = ', '.join(self.c_undefined_value(item)
-                               for item in rtuple.types)
+            undefined_values = []
+            for item in rtuple.types:
+                if not isinstance(item, RTuple):
+                    undefined_values.append(self.c_undefined_value(item))
+                else:
+                    # TODO this needs to be recursive
+                    undefined_values.append(
+                        '{{ {} }}'.format(', '.join(map(self.c_undefined_value, item.types))))
+            values = ', '.join(undefined_values)
             init = 'struct {} {} = {{ {} }};'.format(struct_name, name, values)
             context.statics[name] = init
         return name

--- a/mypyc/emit.py
+++ b/mypyc/emit.py
@@ -176,12 +176,12 @@ class Emitter:
             c_type_compare_val: Callable[[RType], str], compare: str) -> str:
         item_type = rtuple.types[0]
         if not isinstance(item_type, RTuple):
-            return '{}.f0 {} {}'\
-                .format(tuple_expr_in_c, compare, c_type_compare_val(item_type))
+            return '{}.f0 {} {}'.format(
+                tuple_expr_in_c, compare, c_type_compare_val(item_type))
         elif isinstance(item_type, RTuple) and len(item_type.types) == 0:
             # empty tuple
-            return '{}.dummy_var_to_avoid_empty_struct {} {}'\
-                .format(tuple_expr_in_c, compare, c_type_compare_val(int_rprimitive))
+            return '{}.dummy_var_to_avoid_empty_struct {} {}'.format(
+                tuple_expr_in_c, compare, c_type_compare_val(int_rprimitive))
         else:
             return self.tuple_undefined_check_cond(
                 item_type, tuple_expr_in_c + '.f0', c_type_compare_val, compare)

--- a/mypyc/emit.py
+++ b/mypyc/emit.py
@@ -1,7 +1,7 @@
 """Utilities for emitting C code."""
 
 from collections import OrderedDict
-from typing import List, Set, Dict, Optional
+from typing import List, Set, Dict, Optional, List
 
 from mypyc.common import REG_PREFIX, STATIC_PREFIX, TYPE_PREFIX, NATIVE_PREFIX
 from mypyc.ops import (
@@ -176,18 +176,23 @@ class Emitter:
         name = 'tuple_undefined_' + id
         if name not in context.statics:
             struct_name = self.tuple_struct_name(rtuple)
-            undefined_values = []
-            for item in rtuple.types:
-                if not isinstance(item, RTuple):
-                    undefined_values.append(self.c_undefined_value(item))
-                else:
-                    # TODO this needs to be recursive
-                    undefined_values.append(
-                        '{{ {} }}'.format(', '.join(map(self.c_undefined_value, item.types))))
-            values = ', '.join(undefined_values)
-            init = 'struct {} {} = {{ {} }};'.format(struct_name, name, values)
+            values = self.tuple_undefined_value_helper(rtuple)
+            init = 'struct {} {} = {{ {} }};'.format(struct_name, name, ''.join(values))
             context.statics[name] = init
         return name
+
+    def tuple_undefined_value_helper(self, rtuple: RTuple) -> List[str]:
+        res = []
+        for item in rtuple.types:
+            if not isinstance(item, RTuple):
+                res.append(self.c_undefined_value(item))
+            else:
+                sub_list = self.tuple_undefined_value_helper(item)
+                res.append('{ ')
+                res.extend(sub_list)
+                res.append(' }')
+            res.append(', ')
+        return res[:-1]
 
     # Higher-level operations
 

--- a/mypyc/emit.py
+++ b/mypyc/emit.py
@@ -171,11 +171,8 @@ class Emitter:
         return result
 
     def tuple_check_cond(
-            self,
-            rtuple: RTuple,
-            tuple_expr_in_c: str,
-            c_compare_type_func: Callable[[RType], str],
-            compare: str) -> str:
+            self, rtuple: RTuple, tuple_expr_in_c: str,
+            c_compare_type_func: Callable[[RType], str], compare: str) -> str:
         item_expr_in_c = tuple_expr_in_c + '.f0'
         item_type = rtuple.types[0]
         while isinstance(item_type, RTuple):

--- a/mypyc/emitclass.py
+++ b/mypyc/emitclass.py
@@ -199,7 +199,7 @@ def generate_native_getters_and_setters(cl: ClassIR,
                 attr_expr = 'self->{}'.format(attr)
                 emitter.emit_line(
                     'if ({}) {{'.format(
-                        emitter.tuple_check_cond(
+                        emitter.tuple_undefined_check_cond(
                             rtype, attr_expr, emitter.c_undefined_value, '==')))
             else:
                 emitter.emit_line(
@@ -224,7 +224,7 @@ def generate_native_getters_and_setters(cl: ClassIR,
                 attr_expr = 'self->{}'.format(attr)
                 emitter.emit_line(
                     'if ({}) {{'
-                    .format(emitter.tuple_check_cond(
+                    .format(emitter.tuple_undefined_check_cond(
                         rtype, attr_expr, emitter.c_undefined_value, "!=")))
             else:
                 emitter.emit_line('if (self->{} != {}) {{'
@@ -485,7 +485,8 @@ def generate_getter(cl: ClassIR,
         attr_expr = 'self->{}'.format(attr)
         emitter.emit_line(
             'if ({}) {{'.format(
-                emitter.tuple_check_cond(rtype, attr_expr, emitter.c_undefined_value, '==')))
+                emitter.tuple_undefined_check_cond(
+                    rtype, attr_expr, emitter.c_undefined_value, '==')))
     else:
         emitter.emit_line('if (self->{} == {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
     emitter.emit_line('PyErr_SetString(PyExc_AttributeError,')
@@ -514,7 +515,8 @@ def generate_setter(cl: ClassIR,
             emitter.emit_line(
                 'if ({}) {{'
                 .format(
-                    emitter.tuple_check_cond(rtype, attr_expr, emitter.c_undefined_value, '!=')))
+                    emitter.tuple_undefined_check_cond(
+                        rtype, attr_expr, emitter.c_undefined_value, '!=')))
         else:
             emitter.emit_line(
                 'if (self->{} != {}) {{'.format(attr, emitter.c_undefined_value(rtype)))

--- a/mypyc/emitclass.py
+++ b/mypyc/emitclass.py
@@ -6,9 +6,7 @@ from typing import Optional, List, Tuple
 
 from mypyc.common import PREFIX, NATIVE_PREFIX, REG_PREFIX, DUNDER_PREFIX
 from mypyc.emit import Emitter
-from mypyc.emitfunc import (
-    native_function_header, generate_tuple_check_code
-)
+from mypyc.emitfunc import native_function_header
 from mypyc.ops import (
     ClassIR, FuncIR, RType, RTuple, Environment, object_rprimitive, FuncSignature,
     VTableMethod, VTableAttr, VTableEntries
@@ -201,7 +199,8 @@ def generate_native_getters_and_setters(cl: ClassIR,
                 attr_expr = 'self->{}'.format(attr)
                 emitter.emit_line(
                     'if ({}) {{'.format(
-                        generate_tuple_check_code(rtype, attr_expr, emitter.c_undefined_value)))
+                        emitter.tuple_check_cond(
+                            rtype, attr_expr, emitter.c_undefined_value, '==')))
             else:
                 emitter.emit_line(
                     'if (self->{} == {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
@@ -225,7 +224,7 @@ def generate_native_getters_and_setters(cl: ClassIR,
                 attr_expr = 'self->{}'.format(attr)
                 emitter.emit_line(
                     'if ({}) {{'
-                    .format(generate_tuple_check_code(
+                    .format(emitter.tuple_check_cond(
                         rtype, attr_expr, emitter.c_undefined_value, "!=")))
             else:
                 emitter.emit_line('if (self->{} != {}) {{'
@@ -486,7 +485,7 @@ def generate_getter(cl: ClassIR,
         attr_expr = 'self->{}'.format(attr)
         emitter.emit_line(
             'if ({}) {{'.format(
-                generate_tuple_check_code(rtype, attr_expr, emitter.c_undefined_value)))
+                emitter.tuple_check_cond(rtype, attr_expr, emitter.c_undefined_value, '==')))
     else:
         emitter.emit_line('if (self->{} == {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
     emitter.emit_line('PyErr_SetString(PyExc_AttributeError,')
@@ -515,7 +514,7 @@ def generate_setter(cl: ClassIR,
             emitter.emit_line(
                 'if ({}) {{'
                 .format(
-                    generate_tuple_check_code(rtype, attr_expr, emitter.c_undefined_value, '!=')))
+                    emitter.tuple_check_cond(rtype, attr_expr, emitter.c_undefined_value, '!=')))
         else:
             emitter.emit_line(
                 'if (self->{} != {}) {{'.format(attr, emitter.c_undefined_value(rtype)))

--- a/mypyc/emitclass.py
+++ b/mypyc/emitclass.py
@@ -6,10 +6,12 @@ from typing import Optional, List, Tuple
 
 from mypyc.common import PREFIX, NATIVE_PREFIX, REG_PREFIX, DUNDER_PREFIX
 from mypyc.emit import Emitter
-from mypyc.emitfunc import native_function_header
+from mypyc.emitfunc import (
+    native_function_header, generate_tuple_check_code
+)
 from mypyc.ops import (
-    ClassIR, FuncIR, RType, Environment, object_rprimitive, FuncSignature,
-    VTableMethod, VTableAttr, VTableEntries,
+    ClassIR, FuncIR, RType, RTuple, Environment, object_rprimitive, FuncSignature,
+    VTableMethod, VTableAttr, VTableEntries
 )
 from mypyc.sametype import is_same_type
 from mypyc.namegen import NameGenerator
@@ -191,12 +193,19 @@ def generate_native_getters_and_setters(cl: ClassIR,
     for attr, rtype in cl.attributes.items():
         # Native getter
         emitter.emit_line('{}{}({} *self)'.format(emitter.ctype_spaced(rtype),
-                                               native_getter_name(cl, attr, emitter.names),
-                                               cl.struct_name(emitter.names)))
+                                                  native_getter_name(cl, attr, emitter.names),
+                                                  cl.struct_name(emitter.names)))
         emitter.emit_line('{')
         if rtype.is_refcounted:
+            if isinstance(rtype, RTuple):
+                attr_expr = 'self->{}'.format(attr)
+                emitter.emit_line(
+                    'if ({}) {{'.format(
+                        generate_tuple_check_code(rtype, attr_expr, emitter.c_undefined_value)))
+            else:
+                emitter.emit_line(
+                    'if (self->{} == {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
             emitter.emit_lines(
-                'if (self->{} == {}) {{'.format(attr, emitter.c_undefined_value(rtype)),
                 'PyErr_SetString(PyExc_AttributeError, "attribute {} of {} undefined");'.format(
                     repr(attr), repr(cl.name)),
                 '} else {')
@@ -205,7 +214,6 @@ def generate_native_getters_and_setters(cl: ClassIR,
         emitter.emit_line('return self->{};'.format(attr))
         emitter.emit_line('}')
         emitter.emit_line()
-
         # Native setter
         emitter.emit_line(
             'bool {}({} *self, {}value)'.format(native_setter_name(cl, attr, emitter.names),
@@ -213,8 +221,15 @@ def generate_native_getters_and_setters(cl: ClassIR,
                                                 emitter.ctype_spaced(rtype)))
         emitter.emit_line('{')
         if rtype.is_refcounted:
-            emitter.emit_line('if (self->{} != {}) {{'.format(attr,
-                                                              emitter.c_undefined_value(rtype)))
+            if isinstance(rtype, RTuple):
+                attr_expr = 'self->{}'.format(attr)
+                emitter.emit_line(
+                    'if ({}) {{'
+                    .format(generate_tuple_check_code(
+                        rtype, attr_expr, emitter.c_undefined_value, "!=")))
+            else:
+                emitter.emit_line('if (self->{} != {}) {{'
+                                  .format(attr, emitter.c_undefined_value(rtype)))
             emitter.emit_dec_ref('self->{}'.format(attr), rtype)
             emitter.emit_line('}')
         emitter.emit_inc_ref('value'.format(attr), rtype)
@@ -430,7 +445,6 @@ def generate_getseter_declarations(cl: ClassIR, emitter: Emitter) -> None:
 def generate_getseters_table(cl: ClassIR,
                              name: str,
                              emitter: Emitter) -> None:
-
     emitter.emit_line('static PyGetSetDef {}[] = {{'.format(name))
     if not cl.is_trait:
         for attr in cl.attributes:
@@ -468,7 +482,13 @@ def generate_getter(cl: ClassIR,
     emitter.emit_line('{}({} *self, void *closure)'.format(getter_name(cl, attr, emitter.names),
                                                            cl.struct_name(emitter.names)))
     emitter.emit_line('{')
-    emitter.emit_line('if (self->{} == {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
+    if isinstance(rtype, RTuple):
+        attr_expr = 'self->{}'.format(attr)
+        emitter.emit_line(
+            'if ({}) {{'.format(
+                generate_tuple_check_code(rtype, attr_expr, emitter.c_undefined_value)))
+    else:
+        emitter.emit_line('if (self->{} == {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
     emitter.emit_line('PyErr_SetString(PyExc_AttributeError,')
     emitter.emit_line('    "attribute {} of {} undefined");'.format(repr(attr),
                                                                     repr(cl.name)))
@@ -490,7 +510,15 @@ def generate_setter(cl: ClassIR,
         cl.struct_name(emitter.names)))
     emitter.emit_line('{')
     if rtype.is_refcounted:
-        emitter.emit_line('if (self->{} != {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
+        if isinstance(rtype, RTuple):
+            attr_expr = 'self->{}'.format(attr)
+            emitter.emit_line(
+                'if ({}) {{'
+                .format(
+                    generate_tuple_check_code(rtype, attr_expr, emitter.c_undefined_value, '!=')))
+        else:
+            emitter.emit_line(
+                'if (self->{} != {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
         emitter.emit_dec_ref('self->{}'.format(attr), rtype)
         emitter.emit_line('}')
     emitter.emit_line('if (value != NULL) {')

--- a/mypyc/emitclass.py
+++ b/mypyc/emitclass.py
@@ -195,15 +195,7 @@ def generate_native_getters_and_setters(cl: ClassIR,
                                                   cl.struct_name(emitter.names)))
         emitter.emit_line('{')
         if rtype.is_refcounted:
-            if isinstance(rtype, RTuple):
-                attr_expr = 'self->{}'.format(attr)
-                emitter.emit_line(
-                    'if ({}) {{'.format(
-                        emitter.tuple_undefined_check_cond(
-                            rtype, attr_expr, emitter.c_undefined_value, '==')))
-            else:
-                emitter.emit_line(
-                    'if (self->{} == {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
+            emit_undefined_check(rtype, emitter, attr, '==')
             emitter.emit_lines(
                 'PyErr_SetString(PyExc_AttributeError, "attribute {} of {} undefined");'.format(
                     repr(attr), repr(cl.name)),
@@ -220,15 +212,7 @@ def generate_native_getters_and_setters(cl: ClassIR,
                                                 emitter.ctype_spaced(rtype)))
         emitter.emit_line('{')
         if rtype.is_refcounted:
-            if isinstance(rtype, RTuple):
-                attr_expr = 'self->{}'.format(attr)
-                emitter.emit_line(
-                    'if ({}) {{'
-                    .format(emitter.tuple_undefined_check_cond(
-                        rtype, attr_expr, emitter.c_undefined_value, "!=")))
-            else:
-                emitter.emit_line('if (self->{} != {}) {{'
-                                  .format(attr, emitter.c_undefined_value(rtype)))
+            emit_undefined_check(rtype, emitter, attr, '!=')
             emitter.emit_dec_ref('self->{}'.format(attr), rtype)
             emitter.emit_line('}')
         emitter.emit_inc_ref('value'.format(attr), rtype)
@@ -481,14 +465,7 @@ def generate_getter(cl: ClassIR,
     emitter.emit_line('{}({} *self, void *closure)'.format(getter_name(cl, attr, emitter.names),
                                                            cl.struct_name(emitter.names)))
     emitter.emit_line('{')
-    if isinstance(rtype, RTuple):
-        attr_expr = 'self->{}'.format(attr)
-        emitter.emit_line(
-            'if ({}) {{'.format(
-                emitter.tuple_undefined_check_cond(
-                    rtype, attr_expr, emitter.c_undefined_value, '==')))
-    else:
-        emitter.emit_line('if (self->{} == {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
+    emit_undefined_check(rtype, emitter, attr, '==')
     emitter.emit_line('PyErr_SetString(PyExc_AttributeError,')
     emitter.emit_line('    "attribute {} of {} undefined");'.format(repr(attr),
                                                                     repr(cl.name)))
@@ -510,16 +487,7 @@ def generate_setter(cl: ClassIR,
         cl.struct_name(emitter.names)))
     emitter.emit_line('{')
     if rtype.is_refcounted:
-        if isinstance(rtype, RTuple):
-            attr_expr = 'self->{}'.format(attr)
-            emitter.emit_line(
-                'if ({}) {{'
-                .format(
-                    emitter.tuple_undefined_check_cond(
-                        rtype, attr_expr, emitter.c_undefined_value, '!=')))
-        else:
-            emitter.emit_line(
-                'if (self->{} != {}) {{'.format(attr, emitter.c_undefined_value(rtype)))
+        emit_undefined_check(rtype, emitter, attr, '!=')
         emitter.emit_dec_ref('self->{}'.format(attr), rtype)
         emitter.emit_line('}')
     emitter.emit_line('if (value != NULL) {')
@@ -557,3 +525,14 @@ def generate_readonly_getter(cl: ClassIR,
         emitter.emit_line('return {}{}((PyObject *) self);'.format(NATIVE_PREFIX,
                                                                    func_ir.cname(emitter.names)))
     emitter.emit_line('}')
+
+def emit_undefined_check(rtype: RType, emitter: Emitter, attr: str, compare: str) ->None:
+    if isinstance(rtype, RTuple):
+        attr_expr = 'self->{}'.format(attr)
+        emitter.emit_line(
+            'if ({}) {{'.format(
+                emitter.tuple_undefined_check_cond(
+                    rtype, attr_expr, emitter.c_undefined_value, compare)))
+    else:
+        emitter.emit_line(
+            'if (self->{} {} {}) {{'.format(attr, compare, emitter.c_undefined_value(rtype)))

--- a/mypyc/emitclass.py
+++ b/mypyc/emitclass.py
@@ -526,6 +526,7 @@ def generate_readonly_getter(cl: ClassIR,
                                                                    func_ir.cname(emitter.names)))
     emitter.emit_line('}')
 
+
 def emit_undefined_check(rtype: RType, emitter: Emitter, attr: str, compare: str) ->None:
     if isinstance(rtype, RTuple):
         attr_expr = 'self->{}'.format(attr)

--- a/mypyc/emitfunc.py
+++ b/mypyc/emitfunc.py
@@ -11,7 +11,6 @@ from mypyc.ops import (
     EmitterInterface, Unreachable, is_int_rprimitive, NAMESPACE_STATIC, NAMESPACE_TYPE,
     RaiseStandardError,
 )
-from typing import Callable
 from mypyc.namegen import NameGenerator
 
 

--- a/mypyc/emitfunc.py
+++ b/mypyc/emitfunc.py
@@ -15,23 +15,6 @@ from typing import Callable
 from mypyc.namegen import NameGenerator
 
 
-def generate_tuple_check_code(
-        rtuple: RTuple,
-        tuple_expr_in_c: str,
-        c_compare_type_func: Callable[[RType], str],
-        compare: str = '==') -> str:
-    item_expr_in_c = tuple_expr_in_c + '.f0'
-    item_type = rtuple.types[0]
-    while isinstance(item_type, RTuple):
-        item_type = item_type.types[0]
-        item_expr_in_c += '.f0'
-    return '{} {} {}'.format(
-        item_expr_in_c,
-        compare,
-        c_compare_type_func(item_type)
-    )
-
-
 def native_function_type(fn: FuncIR, emitter: Emitter) -> str:
     args = ', '.join(emitter.ctype(arg.type) for arg in fn.args)
     ret = emitter.ctype(fn.ret_type)
@@ -118,7 +101,7 @@ class FunctionEmitterVisitor(OpVisitor[None], EmitterInterface):
             compare = '!=' if op.negated else '=='
             if isinstance(typ, RTuple):
                 # TODO: What about empty tuple?
-                cond = generate_tuple_check_code(typ,
+                cond = self.emitter.tuple_check_cond(typ,
                                             self.reg(op.left),
                                             self.c_error_value,
                                             compare)

--- a/mypyc/emitfunc.py
+++ b/mypyc/emitfunc.py
@@ -101,10 +101,10 @@ class FunctionEmitterVisitor(OpVisitor[None], EmitterInterface):
             compare = '!=' if op.negated else '=='
             if isinstance(typ, RTuple):
                 # TODO: What about empty tuple?
-                cond = self.emitter.tuple_check_cond(typ,
-                                            self.reg(op.left),
-                                            self.c_error_value,
-                                            compare)
+                cond = self.emitter.tuple_undefined_check_cond(typ,
+                                                               self.reg(op.left),
+                                                               self.c_error_value,
+                                                               compare)
             else:
                 cond = '{} {} {}'.format(self.reg(op.left),
                                          compare,

--- a/mypyc/emitfunc.py
+++ b/mypyc/emitfunc.py
@@ -11,7 +11,25 @@ from mypyc.ops import (
     EmitterInterface, Unreachable, is_int_rprimitive, NAMESPACE_STATIC, NAMESPACE_TYPE,
     RaiseStandardError,
 )
+from typing import Callable
 from mypyc.namegen import NameGenerator
+
+
+def generate_tuple_check_code(
+        rtuple: RTuple,
+        tuple_expr_in_c: str,
+        c_compare_type_func: Callable[[RType], str],
+        compare: str = '==') -> str:
+    item_expr_in_c = tuple_expr_in_c + '.f0'
+    item_type = rtuple.types[0]
+    while isinstance(item_type, RTuple):
+        item_type = item_type.types[0]
+        item_expr_in_c += '.f0'
+    return '{} {} {}'.format(
+        item_expr_in_c,
+        compare,
+        c_compare_type_func(item_type)
+    )
 
 
 def native_function_type(fn: FuncIR, emitter: Emitter) -> str:
@@ -100,10 +118,10 @@ class FunctionEmitterVisitor(OpVisitor[None], EmitterInterface):
             compare = '!=' if op.negated else '=='
             if isinstance(typ, RTuple):
                 # TODO: What about empty tuple?
-                item_type = typ.types[0]
-                cond = '{}.f0 {} {}'.format(self.reg(op.left),
-                                            compare,
-                                            self.c_error_value(item_type))
+                cond = generate_tuple_check_code(typ,
+                                            self.reg(op.left),
+                                            self.c_error_value,
+                                            compare)
             else:
                 cond = '{} {} {}'.format(self.reg(op.left),
                                          compare,

--- a/test-data/run.test
+++ b/test-data/run.test
@@ -2277,7 +2277,6 @@ def g() -> None:
     except Exception:
         print("caught the exception")
 
-
 def h() -> Tuple[Tuple[Tuple[int, int], int], int, str, object]:
     raise Exception('Intentional exception')
 [file driver.py]

--- a/test-data/run.test
+++ b/test-data/run.test
@@ -2268,14 +2268,14 @@ class C:
     b: Tuple[Tuple[Tuple[int, int], int], int, str, object]
 def f() -> None:
     c = C()
-    c.b = (((1, 2), 2), 1, "hi", "hi2")
+    c.b = (((1, 2), 2), 1, 'hi', 'hi2')
     print(c.b)
 
 def g() -> None:
     try:
         h()
     except Exception:
-        print("caught the exception")
+        print('caught the exception')
 
 def h() -> Tuple[Tuple[Tuple[int, int], int], int, str, object]:
     raise Exception('Intentional exception')

--- a/test-data/run.test
+++ b/test-data/run.test
@@ -2265,6 +2265,13 @@ Lol(a=1, b=[])
 [case testTupleAttr]
 from typing import Tuple
 class C:
-    d: Tuple[Tuple[int, int], int, str, object]
+    a: int
+    b: Tuple[Tuple[Tuple[int, int], int], int, str, object]
+def f() -> None:
+    c = C()
+    c.a = 1
+    c.b = (((1, 2), 2), 1, "hi", "hi2")
 [file driver.py]
+from native import f
+f()
 [out]

--- a/test-data/run.test
+++ b/test-data/run.test
@@ -2261,3 +2261,10 @@ typing.List[native.A] typing.List[typing.Tuple[int, str]] typing.List[typing.Lis
 Lol(a=1, b=[])
 10
 1
+
+[case testTupleAttr]
+from typing import Tuple
+class C:
+    d: Tuple[Tuple[int, int], int, str, object]
+[file driver.py]
+[out]

--- a/test-data/run.test
+++ b/test-data/run.test
@@ -2271,7 +2271,23 @@ def f() -> None:
     c = C()
     c.a = 1
     c.b = (((1, 2), 2), 1, "hi", "hi2")
+    print(c.a)
+    print(c.b)
+
+def g() -> None:
+    try:
+        h()
+    except Exception:
+        print("caught the exception")
+
+
+def h() -> Tuple[Tuple[Tuple[int, int], int], int, str, object]:
+    raise Exception('Intentional exception')
 [file driver.py]
-from native import f
+from native import f, g
 f()
+g()
 [out]
+1
+(((1, 2), 2), 1, 'hi', 'hi2')
+caught the exception

--- a/test-data/run.test
+++ b/test-data/run.test
@@ -2265,13 +2265,10 @@ Lol(a=1, b=[])
 [case testTupleAttr]
 from typing import Tuple
 class C:
-    a: int
     b: Tuple[Tuple[Tuple[int, int], int], int, str, object]
 def f() -> None:
     c = C()
-    c.a = 1
     c.b = (((1, 2), 2), 1, "hi", "hi2")
-    print(c.a)
     print(c.b)
 
 def g() -> None:
@@ -2288,6 +2285,5 @@ from native import f, g
 f()
 g()
 [out]
-1
 (((1, 2), 2), 1, 'hi', 'hi2')
 caught the exception

